external/kafka-0-10-sql/src/main/scala/org/apache/spark/sql/kafka010/CachedKafkaProducer.scala

```diff
@@ -86,27 +112,38 @@ private[kafka010] object CachedKafkaProducer extends Logging {
     paramsSeq
   }
 
-  /** For explicitly closing kafka producer */
+  /** For explicitly closing kafka producer, will not close an inuse producer until released. */
   private[kafka010] def close(kafkaParams: ju.Map[String, Object]): Unit = {
     val paramsSeq = paramsToSeq(kafkaParams)
     guavaCache.invalidate(paramsSeq)
   }
 
-  /** Auto close on cache evict */
-  private def close(paramsSeq: Seq[(String, Object)], producer: Producer): Unit = {
+  /** Close this producer and process pending closes. */
+  private def close(producer: CachedKafkaProducer): Unit = synchronized {
```

Comment:
Is there a guarantee that this will be called? I can imagine a scenario where long running tasks make all
producers to be evicted, then at some point ` producer.inUseCount.decrementAndGet()` is called in the write task for each producer and this causes all of them to be closable but the method here is never called.

Reply:
We try to close each time, a producer goes not in use, when not in cache.